### PR TITLE
feat: pywal theme

### PR DESCRIPTION
**Pull request by mtoohey31**
A theme that pulls colors from [dylanaraps/pywal](https://github.com/dylanaraps/pywal).

In terms of screenshots, I'm not really sure what should be done because how it looks depends on the pywal theme... let me know what you think!

### CODE REVIEW

File: lua/lualine/themes/pywal.lua [hunk 1]
```diff
@@ -0,0 +1,36 @@
+local wal_colors_file = io.open(os.getenv("HOME") .. "/.cache/wal/colors.sh", "r")
+
+if wal_colors_file == nil then
+  vim.api.nvim_command('echohl ErrorMsg | echo "lightline.lua: ~/.config/wal/colors.sh could not be read" | echohl None')
```
Comment (shadmansaleh) [inline lua/lualine/themes/pywal.lua:4]: Use `add_notice` in `lualine/utils/notices` show the error. And use `lazy_require`
 like [this](https://github.com/shadmansaleh/lualine.nvim/blob/e409e9fc32bfed93111f2205215bc7b57e4f62d0/lua/lualine.lua#L9)
 also it's not lightline.

File: lua/lualine/themes/pywal.lua [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+  error("")
+end
+
+local colors = {}
+local curr_line = wal_colors_file:read()
```
Comment (shadmansaleh) [inline lua/lualine/themes/pywal.lua:9]: You should probably put this in pcall . Because if read fails for some reason you will exit out of the file woth error and the file descriptor won't be closed. So error should be properly handled

File: lua/lualine/themes/pywal.lua [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+
+while curr_line ~= nil do
+  if curr_line:match('^[a-z0-9]+=\'#[a-f0-9]+\'') then
+    local div = curr_line:find('=')
+    local key = curr_line:sub(0, div - 1)
+    local value = curr_line:sub(div + 2, #curr_line - 1)
+    table.insert(colors, key)
```
Comment (shadmansaleh) [inline lua/lualine/themes/pywal.lua:16]: Why is this needed ?

File: lua/lualine/themes/pywal.lua (hunk 2 above)
Comment (shadmansaleh) [reply lua/lualine/themes/pywal.lua:9]: Actually it'll probably be better if you read the entire file  them use `vim.gsplit` for the loop instead of reading line by line /

**Comment by shadmansaleh** [comment]
I think in THEMES.md you should at least provide a description like `auto` theme . Better if you can provide screenshots from a few different themes by pywal . You won't have to put all different mode just different looks on `normal` mode should be enough.

File: lua/lualine/themes/pywal.lua [hunk 4]
```diff
@@ -0,0 +1,36 @@
+local wal_colors_file = io.open(os.getenv("HOME") .. "/.cache/wal/colors.sh", "r")
```
Comment (shadmansaleh) [inline lua/lualine/themes/pywal.lua:1]: Don't hardcode separators ('/') . Use `package.config:sub(1,1)` for separator.

Comment (shadmansaleh) [reply lua/lualine/themes/pywal.lua:1]: Actuslly what OS does pywal support ?

File: lua/lualine/themes/pywal.lua (hunk 1 above)
Comment (mtoohey31) [reply lua/lualine/themes/pywal.lua:4]: Whoops, my bad, fixed in 8fca1e78

File: lua/lualine/themes/pywal.lua (hunk 3 above)
Comment (mtoohey31) [reply lua/lualine/themes/pywal.lua:16]: This adds the variables/colors to the colors object as it reads the file, there might be a more elegant way to do this though, if you have ideas feel free to make suggestions

**Comment by mtoohey31** [comment]
> I think in THEMES.md you should at least provide a description like `auto` theme . Better if you can provide screenshots from a few different themes by pywal . You won't have to put all different mode just different looks on `normal` mode should be enough.

Will do, also, I seem to have closed the PR when I force pushed to fix the history. There doesn't appear to be any way to reopen this PR, but I'll get things fixed in here then open a new one.

File: lua/lualine/themes/pywal.lua (hunk 4 above)
Comment (mtoohey31) [reply lua/lualine/themes/pywal.lua:1]: It supports linux, macos and bsd, and kind of supports windows, so I'll do that.

Comment (mtoohey31) [reply lua/lualine/themes/pywal.lua:1]: Fixed in e7eedaf0

**Comment by shadmansaleh** [comment]
> I force pushed to fix the history

I think you edited history on a shallow clone .

File: lua/lualine/themes/pywal.lua (hunk 3 above)
Comment (shadmansaleh) [reply lua/lualine/themes/pywal.lua:16]: I was asking because you're doing that on [following line](https://github.com/mtoohey31/lualine.nvim/blob/e7eedaf0325a4f009d2c96a6c77aabe483dbd058/lua/lualine/themes/pywal.lua#L23) .

**Comment by mtoohey31** [comment]
I think you're right, I seem to have done that at some point but the current state of my repo isn't a shallow clone any more, I don't think, and it still won't let me reopen it... anyways, I need to get to sleep but I'll finish fixing things tomorrow. Also one thing, would you be able to point me in the right direction for [this comment](https://github.com/shadmansaleh/lualine.nvim/pull/28#discussion_r701543143) with an example of reading a file with pcall? Right now I've got:

```lua
local ok, wal_colors_text = pcall(wal_colors_file.read, '*a')
```

...but that doesn't behave as expected.

**Comment by shadmansaleh** [comment]
You should do

```lua
local ok, wal_colors_text = pcall(io.read, wal_colors_file, '*a')
```

Also instead of [long concats](https://github.com/mtoohey31/lualine.nvim/blob/e7eedaf0325a4f009d2c96a6c77aabe483dbd058/lua/lualine/themes/pywal.lua#L6)

Make a table of parts and and use table.concat . Multiple concations(..) is slow in lua.

**Comment by mtoohey31** [comment]
> Make a table of parts and and use table.concat . Multiple concations(..) is slow in lua.

Will do, as for the `pcall` example though, maybe it's just my version of lua, but that results in `bad argument #1 to '?' (invalid option)`. I'm using `LuaJIT 2.0.5`.

**Comment by shadmansaleh** [comment]
Sorry my example was wrong.

You'll have to do
```lua
local ok, wal_colors_text = pcall(wal_colors_file.read, wal_colors_file, '*a')
```

File: lua/lualine/themes/pywal.lua (hunk 2 above)
Comment (mtoohey31) [reply lua/lualine/themes/pywal.lua:9]: Fixed in b4dd8daa

File: lua/lualine/themes/pywal.lua (hunk 3 above)
Comment (mtoohey31) [reply lua/lualine/themes/pywal.lua:16]: You're totally right, also fixed in b4dd8daa

**Comment by mtoohey31** [comment]
No worries, long concats are fixed in 7bcd5f1

**Comment by mtoohey31** [comment]
So I think it's just documentation and screenshots to go, I'll try and get that done tonight then open a new PR, let me know if there's anything else though!

**Comment by mtoohey31** [comment]
Ok, screenshots are added in c540c7e3, the new PR is open at #35 .
